node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,684 @@
+	fn message_from_validator_needed_by_peer(
+		&self,
+		peer: &PeerId,
+		validator: &ValidatorId,
+	) -> bool {
+		if let Some(set) = self.message_sent_to_peer.get(peer) {
+			!set.contains(validator)
+		} else {
+			false
+		}
+	}
+}
+
+fn network_update_message(n: NetworkBridgeEvent) -> AllMessages {
+	AllMessages::BitfieldDistribution(BitfieldDistributionMessage::NetworkBridgeUpdate(n))
+}
+
+/// The bitfield distribution subsystem.
+pub struct BitfieldDistribution;
+
+impl BitfieldDistribution {
+	/// The protocol identifier for bitfield distribution.
+	const PROTOCOL_ID: ProtocolId = *b"bitd";
+
+	/// Start processing work as passed on from the Overseer.
+	async fn run<Context>(mut ctx: Context) -> SubsystemResult<()>
+	where
+		Context: SubsystemContext<Message = BitfieldDistributionMessage>,
+	{
+		// startup: register the network protocol with the bridge.
+		ctx.send_message(AllMessages::NetworkBridge(
+			NetworkBridgeMessage::RegisterEventProducer(Self::PROTOCOL_ID, network_update_message),
+		))
+		.await?;
+
+		// work: process incoming messages from the overseer and process accordingly.
+		let mut tracker = Tracker::default();
+		loop {
+			{
+				let message = ctx.recv().await?;
+				match message {
+					FromOverseer::Communication { msg } => {
+						// another subsystem created this signed availability bitfield messages
+						match msg {
+							// relay_message a bitfield via gossip to other validators
+							BitfieldDistributionMessage::DistributeBitfield(
+								hash,
+								signed_availability,
+							) => {
+								trace!(target: "bitd", "Processing DistributeBitfield");
+								let job_data = &mut tracker.per_relay_parent.get_mut(&hash).expect("Overseer does not send work items related to relay parents that are not part of our workset. qed");
+
+								let validator = {
+									job_data
+										.validator_set
+										.get(signed_availability.validator_index() as usize)
+										.expect("Our own validation index exists. qed")
+								}
+								.clone();
+
+								let peer_views = &mut tracker.peer_views;
+								let msg = BitfieldGossipMessage {
+									relay_parent: hash,
+									signed_availability,
+								};
+
+								relay_message(&mut ctx, job_data, peer_views, validator, msg)
+									.await?;
+							}
+							BitfieldDistributionMessage::NetworkBridgeUpdate(event) => {
+								trace!(target: "bitd", "Processing NetworkMessage");
+								// a network message was received
+								if let Err(e) =
+									handle_network_msg(&mut ctx, &mut tracker, event).await
+								{
+									warn!(target: "bitd", "Failed to handle incomming network messages: {:?}", e);
+								}
+							}
+						}
+					}
+					FromOverseer::Signal(OverseerSignal::StartWork(relay_parent)) => {
+						trace!(target: "bitd", "Start {:?}", relay_parent);
+						// query basic system parameters once
+						// @todo assumption: these cannot change within a session
```

Comment:
the signing context will change, but the validator set won't. Also, we can't commit any `todo`s that aren't linked to a GitHub issue, whereas the ones here seem more like comments for review